tests/ifs/ifs.c

```diff
@@ -82,8 +84,83 @@ static ssize_t read_file(int dfd, const char *filename, char buf[static restrict
         return n;
 }
 
+static ssize_t read_file(int dfd, const char *filename, char buf[static restrict BUFLEN])
+{
+        int fd = openat(dfd, filename, O_RDONLY | O_CLOEXEC);
+        if (fd < 0)
+            return fd;
+
+        return read_file_fd(fd, buf);
+}
+
+static bool load_test_file(int dfd, int batch_fd, struct test *test, ifs_test_t *ifs_info)
+{
+    char current_buf[BUFLEN], status_buf[BUFLEN];
+    int next_test, current_test;
+
+    /* read both files status and current_batch */
+    read_file(dfd, "status", status_buf);
+    read_file_fd(batch_fd, current_buf);
+
+    /* when previous run has a status of fail, skip test */
+    if (memcmp(status_buf, "fail", strlen("fail")) == 0)
+    {
+        log_warning("Previous run failure found! Refusing to run");
+        return false;
+    }
+
+    /* get interactive test file if provided by user */
+    next_test = get_testspecific_knob_value_uint(test, "test_file", -1);
+    if (next_test == -1)
+    {
+        if (memcmp(current_buf, "none", strlen("none")) == 0)
+            next_test = DEFAULT_TEST_ID;
+        else
+        {
+            current_test = (int) strtoul(current_buf, NULL, 0);
+            if (current_test < 0 && errno == ERANGE)
+            {
+                log_info("Cannot parse current_batch value: %s", current_buf);
+                return false;
+            }
+
+            if (memcmp(status_buf, "untested", strlen("untested")) == 0)
+            {
+                log_info("Test file %s remains untested, so try again", current_buf);
+                next_test = current_test;
+            }
+            else
+                next_test = current_test + 1;
+        }
+    }
+
+    /* write next test file ID */
+    sprintf(ifs_info->image_id, "%#x", next_test);
+    if (write_file(dfd, "current_batch", ifs_info->image_id))
```

Comment:
The call to `read_file_fd` closed the `batch_fd`, but you could have kept it open and used it here.

Reply:
hahaha "I am one with the Force and the Force is with me".

I added the "enforce_run" knob so we can run, even if previous run fails.
Regarding to the write, I'd like to leave it for later and revamp the whole test.